### ISSUE TITLE
Question: can git-bug bugs "branch" and "merge" (similar/same as regular git files/content)?

### ISSUE THREAD

**Issue by johnnyutahh** [MichaelMure/git-bug#16]
Question: can git-bug bugs "branch" and "merge" (similar/same as regular git files/content)?

If not, does it make sense to add such a feature in the future?

**Comment by MichaelMure** [comment]
Hmm, I'm not sure what you have in mind. Could you explain in more details your usecase ?

**Comment by johnnyutahh** [comment]
Here's my first attempt at the thought, which is designed more to catalyze others' thought process, and possibly less to exactly specify:

Document, requirements, bugs, feature requests: they can be context-sensitive to branches just like source code is. So, as an illustration, can a bug follow and "adapt" to a branch same as the source-code does?

A possibly different, additional feature:
Most all other systems require a bug to manually track all the related branches (to the bug). It might be nice for a "built-in" (like git-bug) system to do this implicitly, or some flavor of "automatically."

**Comment by elefthei** [comment]
@johnnyutahh I don't think it makes much sense to make bugs branch sensitive. A bug is a bug on the "master" or "production" version of the repo, however hyperlinks to other branches should be available à-la-github.

As a counter argument, a bug-fix can have its own branch and pull request, in which case you'd have to chose between duplicating bugs to the new bugfix branch or keeping them in master. I say don't duplicate, and bugs refer to the project at large, not a specific branch. This also has the nice side effect of discouraging parallel, divergent, version branches.

**Comment by yrashk** [comment]
In the data model of SIT (a somewhat similar project to this one), making issues branchable was one of the original "must have features". Being designed around a simpler primitive of a file set (instead of depending on Git or any other system), it's all just files, potentially checked into a repo. Once they are checked in, they are branched at will and there are no merge conflicts because all files are considered immutable.

I personally find this to be a very useful feature for back-porting, distinguishing between production/master and particularly pinpointing the state of an issue to any revision ("how did it look *exactly* here?", "is this fixed on 1.0 branch?", etc.)

**Comment by gedw99** [comment]
This discussion raises two really important questions for me.

There are two wishes and their own use cases backing them.

1. "branch insensitive". - You want bugs ( which is a type of document) to be global ( branch independent) to a repo and so reference issues across different paths. 
@MichaelMure  is talking about this.

2 "branch sensitive". - You want documents to live in branches.
@yrashk is talking this this

--

I can see both being really useful. Branch Sensitive expands the use cases to allow the Git repo to be used as a Document Store and Synchronisation system, which is pretty cool. THis whole area is being explored using CRT, OT, etc etc and is really had to get right.  Its a hugely valuable thing to me i admit because it really allows expansion of the use cases.

I dont want to push either, because git-bug has its focus and Branch sensitive is expanding the scope.
Also Branch sensitive ONLY works with no potential merge conflicts because its uses OT techniques. This allows merges from one branch into another and with the lamport clocks allows ordering and so removes potential conflicts. 
@yrashk A very neat trick and i gotta say good stuff as lots of people being trying hard to get that working well.

**Comment by MichaelMure** [comment]
For the record, Lamport clock are used in git-bug only to order bugs between each other. There is no clock involved inside a single bug. The ordering of the edit operations is fixed by the chain of commit and the position of each operation in the array of OperationPack.

**Comment by berkus** [comment]
Just to confirm — current version of git-bug does NOT track bugs differently in different branches bug state is thus global to the repository?

**Comment by MichaelMure** [comment]
@berkus That's correct. The data model and the storage is entirely decoupled from the normal code.

But it's still possible to make git-bug understand what a branch is and make it track where and how a bug is resolved.

It would be interesting to see which bug trackers have this feature and how they handle this problem.

**Comment by yrashk** [comment]
As for other trackers, as I mentioned before, SIT is not only branch-aware but revision-aware as its  shares the namespace with the normal files (when used in conjunction with SCM). This has its pros and cons. Main con is that it exposes these files and their changes to the users. Some pros are: being able to pinpoint issue state, being able to affect issue(s) from within patches, easing transition between version control tools, etc.

I believe that issues can't be "global" -- their content and state always have a context. That's why the design choice to make this data live alongside with the primary content was made. 

From what I remember, I don't remember any decentralized (and merge-conflict-free) issue tracker do anything similar. Though I might have missed some.

**Comment by berkus** [comment]
http://www.bugseverywhere.org/ is also tracking bugs in branches and "merges" ticket resolution together with branch merges.

**Comment by MichaelMure** [comment]
@berkus bugseverywhere gain this feature by storing bugs data in regular branches, which is something git-bug try to avoid. The question is rather if and how git-bug should be made aware of the normal coding process.

I see two possibilities:

- an "active" solution: git-bug listen (via hooks) to events in git (commit, merge ...) and trigger bug changes from that
- a "passive" solution: git-bug store extra metadata (for example, what branch is concerned by the bug, what commit fix it), and is able to compute on demand the bug status (is the bug fixed in the current branch ?)

A passive solution would be easier imho, but this needs more though. This is not only a technical question, it's also what feature do we really want here.

**Comment by neithernut** [comment]
FYI: [git-dit](https://github.com/neithernut/git-dit) allows cross-referencing between issues and from one issue to an arbitrary revision using the parent reference of a commit.

For accumulating metadata and other mechanics confined to a single issue, we only use the first parent of a commit. All other parents are considered references to data external to the issue. This allows, for example, to refer to a revision in which an issue was resolved when closing it. We don't have any real tooling for using those references yet, though.

**Comment by muggenhor** [comment]
I believe the main value of the approach taken by issue trackers that store their data alongside the code is that it has the effect of merging in the state changes to bugs along with their fixes.

I think something similar could be achieved by tracking both the bug-introducing commit and each commit that fixes the bug in the bug's metadata. This wouldn't be as automatic as having the bug metadata stored alongside the code. But it would enable automated traceability and discovery of branches that contain the bug (without a fix).

**Comment by devinrsmith** [comment]
First of all - git-bug looks awesome @MichaelMure . I was thinking to myself last night "it would be cool if something like this existed", and I found it today!

In my mind though, I'd like the bug operations to be tied to the source commits. That way, I could be working on fix-branch, mark the bug as "closed", but the bug remains "open" on the master branch until I merge fix-branch in (after CI passes and whatnot).

Most of the issue / bug tracking software I use today *is* modeled the way that git-bug works though - I just think it's suboptimal in a git-first world.

I *think* there might be some clever ways to still keep the refs/bugs data *outside* of the source commits, but achieve the same goals. The refs/bugs tree could mimic the exact structure of the source refs, with merge commits that always tie the refs together. It would probably get tricky with rebasing though...

**Comment by johnnyutahh** [comment]
Per [ghost's comment](https://github.com/MichaelMure/git-bug/issues/16#issuecomment-414034628):

Can some current or future solution (`git-bug` or otherwise) support both modes?

eg: possibly start each issue/bug creation with a "global" or a "branch" context. Or something else, not sure.

(Apologies if this was already covered above or somewhere else.)

**Comment by MichaelMure** [comment]
It's not the focus at the moment, but yes, entirely possible.

**Comment by johnnyutahh** [comment]
@MichaelMure (the presumed leader of this project's effort) - great, thanks for the timely reply.

My proposal:

Unless we want to leave this issue open to find other solutions that are "git branch friendly," let's close this issue for now. My motivation for launching this issue has been addressed.

Separately:

If my team is able at some point in the future, we'll offer some path to provide [this "git-branch friendly option mode" feature](https://github.com/MichaelMure/git-bug/issues/16#issuecomment-532746326) with minimal or zero changes to `git-bug`. For now: pls close this issue, particularly if the `git-hub` project likes to reduce their number of open issues.

@MichaelMure - does that work?

**Comment by MichaelMure** [comment]
I'll keep the issue open, it help curious people to get some context about the development, and i can point them to if they are curious about this subject.

**Comment by johnnyutahh** [comment]
@MichaelMure - that sounds great.

Further, if there's any future thing I can do for your and/or your project now or in the future, please do not hesitate to message me. It seems like you and your team are doing great work here.

[And please do not hesitate to be persistent if you or others are seeking assistance. This communications channels is one among many, and it's hard to keep current with every channel - so if you need to repeat a request for help or influence, please do not take it personally. Add to be clear: my name here is an avatar, and not my real name, of course.]

Warm regards,
~J

**Comment by MichaelMure** [comment]
Thanks for the kind words. The team at the moment is composed of:
- myself, working on git-bug on the side of my regular job
- @A-Hilaly, who joined through the google summer of code program, also working on the side of his normal student commitment
- a few people contributing occasionally

If you want to help, I'd say do as with any free software: make it yours, use it, report bugs of feature request and if you can contribute in the form of code or documentation (very useful !). There is also an [OpenCollective](https://opencollective.com/git-bug) that could be useful someday.  I'm still curious who you actually are though ;)

**Comment by johnnyutahh** [comment]
Checking in after some time away: has anyone thought more about this?

It seems as if there's not much (if any) additional interest (beyond mind) in this feature, at least judging by this issue's activity.

**Comment by johnnyutahh** [comment]
@yrashk - where is this "SIT" project you (I think?) mention. I've been unable to find it.

Separately:

A collection/list of distributed-bug-tracking (often git oriented), I hope it's okay for me to record here, for my reference (apologies if I'm hijacking this issue inappropriately):

https://dist-bugs.branchable.com/software/ (`git-bug` is of course referenced here, [introduced by @MichaelMure back in 2018](https://kitenet.net/pipermail/dist-bugs/2018-August/000119.html)).

The more-interesting (to me) additional pages from above website:
https://dist-bugs.branchable.com
https://dist-bugs.branchable.com/use-cases

**Comment by neithernut** [comment]
> @yrashk - where is this "SIT" project you (I think?) mention. I've been unable to find it.

https://github.com/sit-fyi/sit (also listed on https://dist-bugs.branchable.com/software/ btw.)
